zerver/lib/retention.py

```diff
@@ -502,6 +522,172 @@ def move_messages_to_archive(
+        archived_message_ids.append(message_id)
+
+    return archive_data, archived_message_ids
+
+
+def update_direct_message_group_stats_of_archived_messages(
+    archived_message_ids: list[int],
+    archive_data: dict[int, DirectMessageGroupArchiveData],
+) -> None:
+    if not archive_data:
+        return
+
+    # Convert to set for O(1) lookup performance
+    archived_message_ids_set = set(archived_message_ids)
+    recipient_ids_needing_stat_update = []
+
+    dm_group_ids = list(archive_data.keys())
+    dm_groups = DirectMessageGroup.objects.filter(id__in=dm_group_ids)
+
+    for dm_group in dm_groups:
+        dm_group.total_messages -= archive_data[dm_group.id].archived_message_count
+
+        needs_first_message_update = (
+            dm_group.first_message_id
+            and dm_group.first_message_id in archived_message_ids_set
+            and dm_group.total_messages > 0
+        )
+        needs_last_message_update = (
+            dm_group.last_message_id
+            and dm_group.last_message_id in archived_message_ids_set
+            and dm_group.total_messages > 0
+        )
+
+        if needs_first_message_update or needs_last_message_update:
+            recipient_ids_needing_stat_update.append(archive_data[dm_group.id].recipient_id)
+
+    if recipient_ids_needing_stat_update:
+        stat_records = (
+            Message.objects.select_related("recipient")
+            .exclude(id__in=archived_message_ids_set)
+            .filter(recipient_id__in=recipient_ids_needing_stat_update)
+            .values("recipient__type_id")
+            .annotate(min_message_id=Min("id"), max_message_id=Max("id"))
+            .values_list("recipient__type_id", "min_message_id", "max_message_id")
+        )
+
+        stat_map = {dm_group_id: (min_id, max_id) for dm_group_id, min_id, max_id in stat_records}
+
+        for dm_group in dm_groups:
+            if dm_group.id in stat_map:
+                min_message_id, max_message_id = stat_map[dm_group.id]
+                dm_group.first_message_id = min_message_id
+                dm_group.last_message_id = max_message_id
+
+    bulk_update_direct_message_groups_stats(dm_groups)
+
+
+def update_direct_message_group_stats_of_restored_messages(
+    message_ids: list[int], chunk_size: int = MESSAGE_BATCH_SIZE
+) -> None:
+    # a map from direct message group id to set of restored message ids
+    restore_data: dict[int, set[int]] = {}
+
+    # Process in chunks to avoid overwhelming the database
+    message_ids_head = message_ids
+    while message_ids_head:
+        message_ids_chunk = message_ids_head[0:chunk_size]
+        message_ids_head = message_ids_head[chunk_size:]
+
+        affected_dm_groups = (
+            Message.objects.select_related("recipient")
+            .filter(
+                id__in=message_ids_chunk,
+                is_channel_message=False,
+                recipient__type=Recipient.DIRECT_MESSAGE_GROUP,
```

Comment:
[The same comment.](https://github.com/zulip/zulip/pull/36045/files#r2347033716)